plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcMergeSink.java

```diff
@@ -0,0 +1,309 @@
+    private final String schemaName;
+    private final String tableName;
+    private final int columnCount;
+    private final String mergeRowIdConjuncts;
+    private final List<String> mergeRowIdFieldNames;
+    private final List<Type> mergeRowIdFieldTypes;
+    private final List<Integer> reservedChannelsForUpdate;
+
+    private final ConnectorPageSinkId pageSinkId;
+    private final ConnectorPageSink insertSink;
+    private final ConnectorPageSink updateSink;
+    private final ConnectorPageSink deleteSink;
+
+    public JdbcMergeSink(
+            ConnectorSession session,
+            JdbcClient jdbcClient,
+            RemoteQueryModifier remoteQueryModifier,
+            ConnectorMergeTableHandle mergeHandle,
+            ConnectorPageSinkId pageSinkId)
+    {
+        JdbcMergeTableHandle jdbcMergeTableHandle = (JdbcMergeTableHandle) mergeHandle;
+        JdbcOutputTableHandle jdbcOutputTableHandle = jdbcMergeTableHandle.getOutputTableHandle();
+        this.catalogName = jdbcOutputTableHandle.getCatalogName();
+        this.schemaName = jdbcOutputTableHandle.getSchemaName();
+        this.tableName = jdbcOutputTableHandle.getTableName();
+        this.columnCount = jdbcOutputTableHandle.getColumnNames().size();
+
+        ImmutableList.Builder<String> mergeRowIdFieldNamesBuilder = ImmutableList.builder();
+        ImmutableList.Builder<Type> mergeRowIdFieldTypesBuilder = ImmutableList.builder();
+        RowType mergeRowIdColumnType = (RowType) jdbcMergeTableHandle.mergeRowIdColumnHandle().getColumnType();
+        for (RowType.Field field : mergeRowIdColumnType.getFields()) {
+            checkArgument(field.getName().isPresent(), "Merge row id column field must have name");
+            mergeRowIdFieldNamesBuilder.add(field.getName().get());
+            mergeRowIdFieldTypesBuilder.add(field.getType());
+        }
+        this.mergeRowIdFieldNames = mergeRowIdFieldNamesBuilder.build();
+        this.mergeRowIdFieldTypes = mergeRowIdFieldTypesBuilder.build();
+        verify(!mergeRowIdFieldNames.isEmpty() && mergeRowIdFieldNames.size() == mergeRowIdFieldTypes.size());
+        this.mergeRowIdConjuncts = jdbcClient.buildMergeRowIdConjuncts(session, mergeRowIdFieldNames, mergeRowIdFieldTypes);
+        this.reservedChannelsForUpdate = getReservedChannelsForUpdate(jdbcOutputTableHandle);
+        verify(!reservedChannelsForUpdate.isEmpty(), "Update primary key itself is not supported");
+
+        this.pageSinkId = pageSinkId;
+        this.insertSink = createInsertSink(session, jdbcOutputTableHandle, jdbcClient, pageSinkId, remoteQueryModifier);
+        this.updateSink = createUpdateSink(session, jdbcOutputTableHandle, jdbcClient, pageSinkId, remoteQueryModifier);
+        this.deleteSink = createDeleteSink(session, jdbcMergeTableHandle.getDeleteTableHandle(), jdbcClient, pageSinkId, remoteQueryModifier);
+    }
+
+    private List<Integer> getReservedChannelsForUpdate(JdbcOutputTableHandle outputTableHandle)
```

Comment:
you are calling a method in constructor, the methods is using fields of the object. So it requires specific order of statements in constructor to make sure this method works correctly. Consider using static method here.